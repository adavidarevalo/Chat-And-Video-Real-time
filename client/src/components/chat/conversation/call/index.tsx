import React, { useState } from 'react';
import Ringing from './ringing';
import Header from './header';
import CallArea from './area';
import CallActions from './actions';
import { ICall } from '../../../../pages/home';

interface CallProps {
  call: ICall;
  setCall: React.Dispatch<React.SetStateAction<ICall>>;
  stream: MediaStream | null;
  callAccepted: boolean;
  myVideo: React.RefObject<HTMLVideoElement>;
  userVideo: React.RefObject<HTMLVideoElement>;
  answerCall: () => void;
  endCall: () => void;
  totalSecondsInCall: moment.Duration;
  setTotalSecondsInCall: React.Dispatch<React.SetStateAction<moment.Duration>>;
  show: boolean;
}

export default function Call({
  call,
  setCall,
  callAccepted,
  myVideo,
  userVideo,
  answerCall,
  stream,
  endCall,
  totalSecondsInCall,
  setTotalSecondsInCall,
  show,
}: CallProps) {
  const { receivingCall, callEnded, name } = call;

  const [showCallActions, setShowCallActions] = useState(false);
  const [toggle, setToggle] = useState(false);

  return (
    <>
      <div
        className={`fixed top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 w-[350px] h-[550px] z-10 rounded-2xl overflow-hidden callBg
    ${receivingCall && !callAccepted && 'hidden'}
      `}
        onMouseOver={() => setShowCallActions(true)}
        onMouseOut={() => setShowCallActions(false)}
      >
        <div>
          <div>
            <Header />
            <CallArea
              name={name}
              totalSecondsInCall={totalSecondsInCall}
              setTotalSecondsInCall={setTotalSecondsInCall}
              callAccepted={callAccepted}
            />
            {showCallActions && <CallActions endCall={endCall} />}
          </div>
          <div>
            <div>
              {callAccepted && !callEnded && (
                <video
                  ref={userVideo}
                  playsInline
                  muted
                  autoPlay
                  className={`${!toggle ? 'largeVideoCall' : 'SmallVideoCall'}`}
                  onClick={() => setToggle((prev) => !prev)}
                ></video>
              )}
            </div>
            <div>
              {stream && (
                <video
                  ref={myVideo}
                  playsInline
                  muted
                  autoPlay
                  className={`
                ${toggle ? 'largeVideoCall' : 'SmallVideoCall'}
                 ${showCallActions && 'moveVideoCall'}`}
                  onClick={() => setToggle((prev) => !prev)}
                ></video>
              )}
            </div>
          </div>
        </div>
      </div>
      {receivingCall && !callAccepted && (
        <Ringing
          call={call}
          setCall={setCall}
          answerCall={answerCall}
          endCall={endCall}
        />
      )}
      {!callAccepted && show ? (
        <audio src="/audio/ringing.mp3" autoPlay loop></audio>
      ) : null}
    </>
  );
}
