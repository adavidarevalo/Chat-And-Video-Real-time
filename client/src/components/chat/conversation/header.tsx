import React, { useMemo } from 'react';
import { useSelector } from 'react-redux';
import { DotsIcon, SearchLargeIcon } from '../../../icons';
import _ from 'lodash';
import { AppState } from '../../../redux/store';

interface ChatHeaderProps {
  insOnline: boolean;
}

export default function ChatHeader({ insOnline }: ChatHeaderProps) {
  const { activeConversation, conversationTyping } = useSelector((state: AppState) => state.chat);

  const status = useMemo(() => {
    if (conversationTyping.includes(activeConversation?._id || '')) return 'Typing...';
    if (insOnline) return 'Online';
    return '';
  }, [insOnline, activeConversation, conversationTyping]);

  return (
    <div className="h-[52px] dark:bg-dark_bg_2 flex items-center p16 select-none">
      <div className="w-full flex items-center justify-between">
        <div className="flex items-center gap-x-4">
          <button className="btn">
            <img
              src={activeConversation?.picture}
              alt={`${activeConversation?.name} avatar`}
              className="w-full h-full rounded-full object-cover"
            />
          </button>
          <div className="flex flex-col">
            <h1 className="dark:text-white text-md font-bold">
              {_.capitalize(activeConversation?.name.split(' ')[0])}
            </h1>
            <span className="text-xs dark:text-dark_svg_2">{status}</span>
          </div>
        </div>
        <ul className="flex items-center gap-x-2.5">
          <li>
            <button className="btn">
              <SearchLargeIcon className="dark:fill-dark_svg_1" />
            </button>
          </li>
          <li>
            <button className="btn">
              <DotsIcon className="dark:fill-dark_svg_1" />
            </button>
          </li>
        </ul>
      </div>
    </div>
  );
}
