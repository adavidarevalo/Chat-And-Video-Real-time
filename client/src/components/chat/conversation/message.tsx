import moment from 'moment';
import React from 'react'
import { TriangleIcon } from '../../../icons';
import MessageType from './../../../types/message.type';

interface MessageProps {
  isMe: boolean;
  message: MessageType;
}

export default function Message({ message, isMe }: MessageProps) {
  return (
    <div className={`w-full flex mt-2 space-x-3 max-w-xs ${isMe && 'ml-auto justify-end'}`}>
      <div>
        <div
          className={`relative h-full dark:text-dark_text_1 p-2 rounded-lg min-w-[54px] ${
            isMe ? 'bg-green_3' : 'dark:bg-dark_bg_2'
          }`}>
          <p className="float-left h-full text-sm pb-5">{message.message}</p>
          <span className="absolute right-1.5 text-xs pt-6 bottom-1 text-dark_text_5">
            {moment(message.createdAt).format('HH:mm')}
          </span>
          <span>
            {!isMe && <TriangleIcon className="dark:fill-dark_bg_2 rotate-[60deg] absolute top-[5px] -left-1.5" />}
          </span>
        </div>
      </div>
    </div>
  );
}
