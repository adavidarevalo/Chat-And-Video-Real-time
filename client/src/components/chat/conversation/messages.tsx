import { useEffect, useRef } from 'react';
import { useSelector } from 'react-redux';
import Message from './message';
import { AppState } from '../../../redux/store';
import Typing from './typing';
import FileMessage from './files/file_message';

export default function ChatMessages() {
  const { messages, conversationTyping, activeConversation } = useSelector((state: AppState) => state.chat);
  const { user } = useSelector((state: AppState) => state.user);
  const endRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    scrollToBottom();
  }, [messages]);

  const scrollToBottom = () => {
    if (endRef.current) {
      endRef.current.scrollIntoView({ behavior: 'smooth' });
    }
  };

  return (
    <div className={`mb-[60px] bg-[url("https://cdn.wallpapersafari.com/54/0/HluF7g.jpg")] h-[88vh]`}>
      <div className="scrollbar overflow_scrollbar overflow-auto py-2 px-[5%] h-full">
        {messages.length > 0 &&
          messages.map((message: any) => (
            <>
              {message.files.length > 0 &&
                message.files.map((file: any) => (
                  <FileMessage
                    fileMessage={file}
                    message={message}
                    key={message._id}
                    isMe={user?._id === message.sender._id}
                  />
                ))}
              {message.files.length === 0 && message.message.length > 0 && (
                <Message key={message._id} message={message} isMe={user?._id === message.sender._id} />
              )}
            </>
          ))}
        {conversationTyping.includes(activeConversation?._id || '') && <Typing />}
        <div className={'mt-2'} ref={endRef}></div>
      </div>
    </div>
  );
}
