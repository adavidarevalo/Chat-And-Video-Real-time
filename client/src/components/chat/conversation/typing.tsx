import moment from 'moment';
import React from 'react'
import { TriangleIcon } from '../../../icons';
import MessageType from '../../../types/message.type';
import { BeatLoader } from 'react-spinners';


export default function Typing() {
  return (
    <div className={`w-full flex mt-2 space-x-3 max-w-xs`}>
      <div>
        <div className={`relative h-full dark:text-dark_text_1 p-2 rounded-lg min-w-[54px] dark:bg-dark_bg_2`}>
          <div className="float-left h-full text-sm py-2  px-[5px]">
            <BeatLoader color="#36d7b7" size={10} />
          </div>
          <span>
            <TriangleIcon className="dark:fill-dark_bg_2 rotate-[30deg] absolute top-[0px] -left-1.5" />
          </span>
        </div>
      </div>
    </div>
  );
}
