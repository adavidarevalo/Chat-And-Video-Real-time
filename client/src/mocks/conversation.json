{
  "_id": "64d30783582dd5869d6bc1c1",
  "name": "Jose",
  "picture": "https://res.cloudinary.com/dzmirfc34/image/upload/v1691356100/pwdzlicu8ip2lv7dm55a.jpg",
  "isGroup": false,
  "users": [
    {
      "_id": "64c91858a22823a5886ecf06",
      "name": "David Arevalo",
      "email": "davidarevaloc20@gmail.com",
      "picture": "https://res.cloudinary.com/dzmirfc34/image/upload/v1691356100/pwdzlicu8ip2lv7dm55a.jpg",
      "status": "",
      "createdAt": "2023-08-01T14:36:08.293Z",
      "updatedAt": "2023-08-01T14:36:08.293Z",
      "__v": 0
    },
    {
      "_id": "64c92f264942b0176d1ff6dc",
      "name": "Jose",
      "email": "davidarevaloasd20@gmail.com",
      "picture": "https://res.cloudinary.com/dzmirfc34/image/upload/v1691356100/pwdzlicu8ip2lv7dm55a.jpg",
      "status": "",
      "createdAt": "2023-08-01T16:13:26.644Z",
      "updatedAt": "2023-08-01T16:13:26.644Z",
      "__v": 0
    }
  ],
  "createdAt": "2023-08-09T03:26:59.460Z",
  "updatedAt": "2023-08-16T23:54:55.308Z",
  "__v": 0,
  "latestMessage": {
    "_id": "64dd61cfd2ebce5a69104ec7",
    "sender": {
      "_id": "64c91858a22823a5886ecf06",
      "name": "David Arevalo",
      "email": "davidarevaloc20@gmail.com",
      "picture": "https://res.cloudinary.com/dzmirfc34/image/upload/v1691356100/pwdzlicu8ip2lv7dm55a.jpg",
      "status": ""
    },
    "message": "",
    "conversation": "64d30783582dd5869d6bc1c1",
    "files": [
      {
        "file": {
          "asset_id": "1cce3839c828cefadd1b64cb8be5bd7f",
          "public_id": "oqce4o4nwmklq5ypeiw8.jpeg",
          "version": 1692230095,
          "version_id": "37c6da750343363af9aba986b6713cb9",
          "signature": "45c0ea78794c58ae5cbe25a170a981baee02af43",
          "resource_type": "raw",
          "created_at": "2023-08-16T23:54:55Z",
          "tags": [],
          "bytes": 9904,
          "type": "upload",
          "etag": "a2e3d9f86a67dc1b3d2811b929a6d6e4",
          "placeholder": false,
          "url": "http://res.cloudinary.com/dzmirfc34/raw/upload/v1692230095/oqce4o4nwmklq5ypeiw8.jpeg",
          "secure_url": "https://res.cloudinary.com/dzmirfc34/raw/upload/v1692230095/oqce4o4nwmklq5ypeiw8.jpeg",
          "folder": "",
          "original_filename": "download"
        },
        "type": "IMAGE"
      }
    ],
    "createdAt": "2023-08-16T23:54:55.266Z",
    "updatedAt": "2023-08-16T23:54:55.266Z",
    "__v": 0
  }
}
