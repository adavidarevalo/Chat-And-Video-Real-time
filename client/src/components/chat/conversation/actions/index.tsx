import React, { useRef, useState } from 'react';
import EmojiPicker from './emoji_picker';
import Attachments from './attachments';
import Input from './input';
import { SendIcon } from '../../../../icons';
import { useDispatch, useSelector } from 'react-redux';
import { sendMessage } from '../../../../redux/actions/chat.actions';
import { ClipLoader } from 'react-spinners';
import { AppState } from '../../../../redux/store';
import { useSocket } from '../../../../context/socket.context';

export default function ChatActions() {
  const [message, setMessage] = useState('');
  const [showAttachments, setShowAttachments] = useState(false);
  const [showEmoji, setShowEmoji] = useState(false);
  const [loading, setLoading] = useState(false);
  const { activeConversation, status } = useSelector((state: AppState) => state.chat);
  const { user } = useSelector((state: AppState) => state.user);
  const dispatch = useDispatch();
  const socket = useSocket();
  const textRef = useRef<HTMLInputElement>(null);

  const handlerSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setLoading(true);
    const value = {
      message,
      conversation_id: activeConversation?._id,
      token: user?.token || '',
      files: [],
    };
    const newMessage = await dispatch(sendMessage(value) as any);
    socket?.socket.emit('send message', newMessage.payload);
    setMessage('');
    setLoading(false);
    setShowEmoji(false);
  };
  return (
    <form
      onSubmit={handlerSubmit}
      className="dark:bg-dark_bg_2 h-[60px] w-full flex items-center absolute bottom-0 py-2 px-4 select-none"
    >
      <div className="w-full flex items-center gap-x-2">
        <ul className="flex gap-x-2">
          <EmojiPicker
            textRef={textRef}
            message={message}
            setMessage={setMessage}
            showEmoji={showEmoji}
            setShowEmoji={setShowEmoji}
            showAttachments={showAttachments}
            setShowAttachments={setShowAttachments}
          />
          <Attachments
            showEmoji={showEmoji}
            setShowEmoji={setShowEmoji}
            showAttachments={showAttachments}
            setShowAttachments={setShowAttachments}
          />
        </ul>
        <Input message={message} setMessage={setMessage} textRef={textRef} />
        <button className="btn" type="submit" disabled={status === 'loading'}>
          {status === 'loading' && loading ? (
            <ClipLoader color="#E9EDEF" size={25} />
          ) : (
            <SendIcon className="dark:fill-dark_svg_1" />
          )}
        </button>
      </div>
    </form>
  );
}
