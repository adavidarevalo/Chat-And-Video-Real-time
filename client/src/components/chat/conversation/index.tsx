import React, { useEffect, useMemo } from 'react';
import ChatHeader from './header';
import ChatMessages from './messages';
import { useDispatch, useSelector } from 'react-redux';
import { getConversationMessages } from '../../../redux/actions/chat.actions';
import ChatActions from './actions';
import { AppDispatch, AppState } from '../../../redux/store';
import { getConversationId } from '../../../utils/get_conversation';
import FilesPreview from './preview/files';
import { User } from '../../../types/user.type';

interface ConversationContainerProps {
  callUser: () => void
}

export default function ConversationContainer({ callUser }: ConversationContainerProps) {
  const dispatch = useDispatch<AppDispatch>();

  const { activeConversation, onlineUsers, files } = useSelector(
    (state: AppState) => state.chat,
    );

  const { user } = useSelector((state: AppState) => state.user);

  useEffect(() => {
    if (activeConversation?._id) {
      const values = {
        token: user?.token,
        conversation_id: activeConversation?._id,
      };

    dispatch(getConversationMessages(values));
    }
  }, [activeConversation]);

  const insOnline = useMemo(() => {
    return onlineUsers.some(
      ({ userId }) =>
        userId ===
        getConversationId(user as User, activeConversation?.users as User[]),
    );
  }, [activeConversation, onlineUsers]);

  return (
    <div className="relative h-full w-full dark:bg-dark_bg_4 select-none border-1 dark:border-l-dark_border_1 border-b-green_2 select-none overflow-inherit">
      <div>
        <ChatHeader insOnline={insOnline} callUser={callUser} />
        {files.length > 0 ? (
          <FilesPreview />
        ) : (
          <>
            <ChatMessages />
            <ChatActions />
          </>
        )}
      </div>
    </div>
  );
}
