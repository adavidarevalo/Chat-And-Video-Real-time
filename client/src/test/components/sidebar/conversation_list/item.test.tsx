import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import conversation from './../../../../mocks/conversation.json';
import ConversationItem from '../../../../components/sidebar/conversation_list/item';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import user from './../../chat/conversation/mocks/user.json';
import activeConversation from "./../../../../mocks/active_conversation.json"

const mockStore = configureMockStore([]);

describe('ConversationItem', () => {
  it('should render correctly', () => {
    const customState = {
      user: {
        user,
      },
      chat: {
        activeConversation,
        conversationTyping: [],
      },
    };
    const store = mockStore(customState);

    const { container } = render(
      <Provider store={store}>
        <ConversationItem isOnline={false} conversation={conversation as any} />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByText('Jose')).toBeInTheDocument();
    expect(screen.getByText('16/08/2023')).toBeInTheDocument();
  });

    it('Validate user typing', async () => {
      const customState = {
        user: {
          user,
        },
        chat: {
          activeConversation,
          conversationTyping: ['64d30783582dd5869d6bc1c1'],
        },
      };
      const store = mockStore(customState);

      render(
        <Provider store={store}>
          <ConversationItem
            isOnline={false}
            conversation={conversation as any}
          />
        </Provider>,
      );

      expect(await screen.findByText('Typing...')).toBeInTheDocument();
    });
});
