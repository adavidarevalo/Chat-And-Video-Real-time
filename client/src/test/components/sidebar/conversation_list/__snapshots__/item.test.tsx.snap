// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversationItem should render correctly 1`] = `
<div>
  <li
    class="list-none h-[72px] w-full dark:bg-dark_bg_1 hover:dark:bg-dark_bg_2 dark:text-dark_1 px-[10px] false"
  >
    <div
      class="relative w-full flex items-center justify-between py-[10px]"
    >
      <div
        class="flex items-center gap-x-3 "
      >
        <div
          class="relative w-[80px] rounded-full overflow-hidden false"
        >
          <img
            alt="Jose"
            class="object-cover rounded-full h-full w-[50px]"
            src="https://res.cloudinary.com/dzmirfc34/image/upload/v1691356100/pwdzlicu8ip2lv7dm55a.jpg"
          />
        </div>
        <div
          class="w-full flex flex-col"
        >
          <h1
            class="font-bold flex items-center gap-x-2 dark:text-dark_text_3"
          >
            Jose
          </h1>
          <div>
            <div
              class="flex items-center gap-x-1 "
            >
              <div
                class="flex-1 items-center gap-x-1 dark:text-dark_text_2"
              >
                <p />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex flex-col gap-y-4 items-end text-xs"
      >
        <span
          class="dark:text-dark_text_2"
        >
          16/08/2023
        </span>
      </div>
    </div>
    <div
      class="ml-14 border-b  dark:border-b-dark_bg_2"
    />
  </li>
</div>
`;
